### ISSUE TITLE
Replace VS projects & autotools with CMake?

### ISSUE THREAD

**Issue by FraGag** [losnoco/vgmstream#326]
Are you guys interested in replacing the current Visual Studio project-based and autotools-based build systems with a single build system based on [CMake](https://cmake.org/)? @bnnm mentioned not being familiar with autotools in https://github.com/kode54/vgmstream/issues/317#issuecomment-443882540. CMake can generate Visual Studio projects, and Visual Studio 2017 also supports CMake projects natively.

I could start working on this next weekend if you like.

**Comment by bnnm** [comment]
Also worth pointing out the Arch Linux PKGBUILD [here](https://aur.archlinux.org/packages/vgmstream-kode54-git/) has support for libcelt (impressive, as it's quite annoying to build), by compiling and adding to vgmstream as a solid lib/.a (as opposed to external lib/.so) I think. May be interesting to do something similar and allow including libs directly to vgmstream as .a (or for Windows w/ .lib), if isn't too complex, since adding DLLs is a pain.

**Comment by FraGag** [comment]
I'll definitely use the PKGBUILD as inspiration when making the CMake-based build system. (Ironically, I am using Arch Linux myself, but I've been building a plain vgmstream straight from the repo myself because I didn't know that package existed and added so much to the vanilla Linux build... To be fair, I was doing so before that package was initially published!)

At the moment, the repo has prebuilt binaries for Windows only. If we want to turn all the DLLs into static libraries, then either we replace the prebuilt DLLs with prebuilt static libraries, or we remove the prebuilt binaries completely and replace them with Git submodules or just copies of the source. At the moment, it looks like none of vgmstream's dependencies have a CMake-based build system; either we shell out to the original build system (or the correct one for the host platform if the dependency has more than one build system...), or we write a CMake-based build system for the dependencies too. Using the original build systems doesn't really give any advantage to using CMake in the first place, though. For the libraries that were developed by @kode54 and @Thealexbarney, such as `libatrac9` and `libg719_decode`, the CMake-based build systems should probably be contributed upstream. For third-party libraries such as FFmpeg, we can write a minimalist build system that does only what we need (imagine creating a Visual C++ project manually for that dependency, but in CMake instead). Perhaps once that's done, we'll be able to remove the various `VGM_USE_*` flags and just support every feature on every platform; if system libraries aren't available (or are too old or not compatible), then we can just use our own versions.

@kode54 Do you have any opinion on which CMake version we should target? The latest is 3.13.1, but many Linux distributions tend to lag behind in software updates in order to provide stability. For example, Debian "Stretch" has CMake 3.7.2, and stretch-backports has 3.9.5. CMake provides backwards-compatible behavior using its [policy mechanism](https://gitlab.kitware.com/cmake/community/wikis/doc/cmake/Policies) based on the minimum required version we specify. Ideally, we would _test_ that the build actually works with that exact version. One strategy would be to start with 2.6 (which is the first version to support policies) and bump the requirement up as I stumble upon bugs or missing features.

**Comment by bnnm** [comment]
Here are my considerations against copying code here or always using static libs: [docs](https://github.com/kode54/vgmstream/blob/master/doc/BUILD.md#external-libraries). It'd be best to give the option to use either dynamic .dll/.so or static .lib/.a, preferably per lib (one may want to use a libatrac9 .a but link against FFmpeg's .so).

So I don't think current DLLs should be removed until we have something more solid and configurable.

For libs, I'd like some simplified way to re-build libs as needed from repos. This would be useful too for 64b support, since current DLLs are 32b. It could set custom config, exact repo version (so I don't have to guess if latest upstream is compatible) and all/one lib (since stuff like FFmpeg takes ages to build, and don't want to rebuild every time libatrac9 would update).

Can't CMake call other Make/VS msbuild directly? Maybe we should have some separate/base script first (or another CMake?) that DLs + builds stuff as dynamic/static libs, and copies them to ext_libs or appropriate dirs. Then vgmstream's CMake can check they exist and uses them. Something in the style of the current .ps1 script but more usable/universal.

(also note how our libcelt(s) are built, that's the harder use-case)

**Comment by bnnm** [comment]
I think it's best to give options and let the person building it decide.

When making CMake scripts do note I will **not** mantain any new .c/h lists. I fix the VS ones only b/c they already existed. So if CMake can't get .c automatically it won't be very useful.

**Comment by FraGag** [comment]
> I think it's best to give options and let the person building it decide.

Oh, of course, there will be options to enable or disable each optional dependency. Ideally, we'd test a few combinations in the AppVeyor build (e.g. one build with no optional dependencies, one build for each optional dependency with only that dependency enabled, and one build with all dependencies enabled).

> When making CMake scripts do note I will **not** mantain any new .c/h lists. I fix the VS ones only b/c they already existed. So if CMake can't get .c automatically it won't be very useful.

CMake supports globs, but it's not recommended to use them, because when a new file appears, CMake doesn't automatically trigger a new build. Listing files explicitly is the only method that always works. Plus, with CMake, the list of files to build is just a list of file names:

```cmake
set(HEADERS
  coding/acm_decoder_libacm.h
  coding/coding.h
  coding/ea_mt_decoder_utk.h
  coding/g72x_state.h
  # more file names here...
)

set(SOURCES
  coding/acm_decoder.c
  coding/acm_decoder_decode.c
  coding/acm_decoder_util.c
  coding/adx_decoder.c
  # a lot more file names here...
)
```

If it's _really_ necessary, I could write tools to update the lists in the `CMakeLists.txt` files. However, I would expect that the updated `CMakeLists.txt` contents then be included in the commit that adds or removes the files (contrarily to the `Makefile.autotools.am` files which just have `(auto-updated)` in the repo).

**Comment by CyberBotX** [comment]
This could probably be closed now that I've made CMake scripts and they were just merged in from pull request #370.
